boa/contracts/vvm/vvm_contract.py

```diff
@@ -58,5 +86,267 @@ def deploy(self, *args, env=None):
+        """
+        return VVMEval(code, self, return_type)()
+
+    @cached_property
+    def _storage(self):
+        """
+        Allows access to the storage variables of the contract.
+        Note that this is quite slow, as it requires the complete contract to be
+        recompiled.
+        """
+
+        def storage():
+            return None
+
+        for name, spec in self.compiler_output["layout"]["storage_layout"].items():
+            setattr(storage, name, VVMStorageVariable(name, spec, self))
+        return storage
+
+    @cached_property
+    def internal(self):
+        """
+        Allows access to internal functions of the contract.
+        Note that this is quite slow, as it requires the complete contract to be
+        recompiled.
+        """
+
+        def internal():
+            return None
+
+        result = self._compile_function_metadata()
+        for fn_name, meta in result.items():
+            if meta["visibility"] == "internal":
+                function = VVMInternalFunction(meta, self)
+                setattr(internal, function.name, function)
+        return internal
+
+    def _compile_function_metadata(self) -> dict:
+        """Compiles the contract and returns the function metadata"""
+
+        # todo: move this to vvm?
+        def run_vyper(filename: str) -> dict:
+            stdoutdata, stderrdata, command, proc = vyper_wrapper(
+                vyper_binary=get_executable(self.vyper_version),
+                f="metadata",
+                source_files=[filename],
+            )
+            return json.loads(stdoutdata)["function_info"]
+
+        if self.filename is not None:
+            return run_vyper(self.filename)
+        with NamedTemporaryFile(suffix=".vy") as f:
+            f.write(self.source_code.encode())
+            f.flush()
+            return run_vyper(f.name)
+
+
+class _VVMInternal(ABIFunction):
+    """
+    An ABI function that temporarily changes the bytecode at the contract's address.
+    Subclasses of this class are used to inject code into the contract via the
+    `source_code` property using the vvm, temporarily changing the bytecode
+    at the contract's address.
+    """
+
+    @cached_property
+    def _override_bytecode(self) -> bytes:
+        assert isinstance(self.contract, VVMContract)  # help mypy
+        source = "\n".join((self.contract.source_code, self.source_code))
+        compiled = vvm.compile_source(source, vyper_version=self.contract.vyper_version)
+        return to_bytes(compiled["<stdin>"]["bytecode_runtime"])
+
+    @property
+    def source_code(self):
+        raise NotImplementedError  # to be implemented in subclasses
+
+    def __call__(self, *args, **kwargs):
+        env = self.contract.env
+        assert isinstance(self.contract, VVMContract)  # help mypy
+        balance_before = env.get_balance(env.eoa)
+        env.set_code(self.contract.address, self._override_bytecode)
+        env.set_balance(env.eoa, 10**20)
```

Comment:
what?

Reply:
the eoa needs to have some balance to call the contract... but maybe that should be handled by the users

Reply:
it does?

Reply:
For example eval can change the state

Reply:
I don't think in general you need balance to call a contract though

Reply:
let's try